### PR TITLE
Remote API Compatibility Documentation. First Attempt.

### PR DESCRIPTION
**Pull request by matthewavery**
Applys to ticket #460

Adds Remote API documentation as well as a compatitibility matrix above for all APIs except `Network` and `Volume`. Which are to be added at a later time. 

@sflxn @hickeng

### CODE REVIEW

**Comment by matthewavery** [comment]
I am seeing some format issues right now. I am going to go in and fix those. I think part of it got messed up due to my merge resolution.

File: apiservers/engine/API-COMPATIBILITY.md [hunk 1]
```diff
@@ -1,7 +1,73 @@
...
+## Container API Support
+
+|API|support?|
+|---|---|
+|List Containers|YES|
+|Create a Container|YES|
+|inspect a container| YES|
+|List processes running inside a container|YES|
+|Get Container Logs|YES|
+|Inspect changes on a containers filesystem|YES|
```
Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: I'm not sure inspecting changes on the container filesystem is a priority.  The user scenarios for this functionality is so limited.  We'll eventually do it, but I'm not sure it'll show up in VIC 1.0.

File: apiservers/engine/API-COMPATIBILITY.md [hunk 2]
```diff
@@ -1,7 +1,73 @@
...
+|Export a container|YES - In the future|
+|Get Container stats based on resource usage|__NO__|
```
Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: I would say this is future and not No.  I think from a production standpoint, it's important.

File: apiservers/engine/API-COMPATIBILITY.md [hunk 3]
```diff
@@ -1,7 +1,73 @@
...
+|Start a Container|YES|
+|Stop a container|YES|
+|Restart a Container|YES|
+|kill a container|YES|
+|Update a Container|*Maybe*|
+|Rename a Container|YES|
+|Pause a Container|*maybe*|
+|Unpause a Container|*maybe*|
+|Attach a Container|YES|
+|Attach a Container(websocket)|YES|
```
Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: I'm not sure we'll get around to the websocket case unless it's straightforward.

Comment (matthewavery) [inline apiservers/engine/API-COMPATIBILITY.md]: Marking that one as a maybe.

File: apiservers/engine/API-COMPATIBILITY.md (hunk 2 above)
Comment (matthewavery) [inline apiservers/engine/API-COMPATIBILITY.md]: marked as `YES - in the future`

File: apiservers/engine/API-COMPATIBILITY.md (hunk 1 above)
Comment (matthewavery) [inline apiservers/engine/API-COMPATIBILITY.md]: marked as `YES - in the future`

File: apiservers/engine/API-COMPATIBILITY.md [hunk 4]
```diff
@@ -1,7 +1,73 @@
...
+|wait a container|*maybe*|
```
Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: I think we may have to implement this after all, given the use of it by docker attach.

File: apiservers/engine/API-COMPATIBILITY.md [hunk 5]
```diff
@@ -1,7 +1,73 @@
...
+|Retrieve information about files and folders in a container|YES|
+|Get and archive of a filesystem resource in a container|YES|
+|Extract an archive of files or folders to a driectory in a container|YES|
+
+## Image API Support
+
+|API|support?|
+|---|---|
+|List Images|YES|
+|Build image from a Dockerfile|YES - In the Future|
```
Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: Let's shorten this to just 'Future'.

File: apiservers/engine/API-COMPATIBILITY.md [hunk 6]
```diff
@@ -1,7 +1,72 @@
 # Docker Remote API Compatibility
 VIC attempts to be compatible with the Docker remote API; however, there are some differences.  The API differences are documented below, grouped together by functional areas.
 
-## Container APIs
+# support Matrix
```
Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: Capitalize s.

File: apiservers/engine/API-COMPATIBILITY.md [hunk 7]
```diff
@@ -10,44 +75,695 @@ GET /containers/json
...
+### Json Fields
+
+**misc**
+
+|field|description|supported?|
+|---|---|---|
+|AttachStdin|Attach to Stdin|YES|
+|AttachStdout|Attach to STDOUT|YES|
+|AttachStderr|Attach to STDERR|YES|
+|Cmd|Array of strings which specify commands to run|*maybe*|
```
Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: YES

File: apiservers/engine/API-COMPATIBILITY.md [hunk 8]
```diff
@@ -10,44 +75,695 @@ GET /containers/json
...
+|Domainname|string to use for the containers domain|YES|
+|Env|Set environment variables|YES|
+|Entrypoint=""|Overwrite the default ENTRYPOINT of the image|YES|
+|Hostname|Container host name|YES|
+|Image|Image name to be used for the container|*maybe*|
+|OpenStdin|Keep STDIN open even if not attached|*maybe*|
+|StdinOnce|close stdin once single attached client disconnects|*maybe*|
```
Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: OpenStdin and StdinOnce are used by attached so let's say YES for now.

File: apiservers/engine/API-COMPATIBILITY.md (hunk 5 above)
Comment (matthewavery) [inline apiservers/engine/API-COMPATIBILITY.md]: Changed

File: apiservers/engine/API-COMPATIBILITY.md [hunk 9]
```diff
@@ -10,44 +75,695 @@ GET /containers/json
...
+
+* 200 : no error
+* 404 : no such container
+* 500 : no error
+
+## List processes running inside a container
+
+```GET containers/(id)/top```
+
+__Not Supported By VIC__
```
Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: Future support.

File: apiservers/engine/API-COMPATIBILITY.md (hunk 6 above)
Comment (matthewavery) [inline apiservers/engine/API-COMPATIBILITY.md]: Done

File: apiservers/engine/API-COMPATIBILITY.md (hunk 7 above)
Comment (matthewavery) [inline apiservers/engine/API-COMPATIBILITY.md]: Done :)

File: apiservers/engine/API-COMPATIBILITY.md (hunk 8 above)
Comment (matthewavery) [inline apiservers/engine/API-COMPATIBILITY.md]: Done

File: apiservers/engine/API-COMPATIBILITY.md (hunk 9 above)
Comment (matthewavery) [inline apiservers/engine/API-COMPATIBILITY.md]: Done.

Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: Let's say Future support instead of not supported.

Comment (matthewavery) [inline apiservers/engine/API-COMPATIBILITY.md]: I just pushed up a new version marking this as `__SUPPORTED in the Future__.

File: apiservers/engine/API-COMPATIBILITY.md [hunk 10]
```diff
@@ -10,44 +76,694 @@ GET /containers/json
...
+### Request
+The request will have one paramter in it's json object.
+
+* resource : the file or folder you want to copy from docker.
+
+### Response
+
+The response will return a tar stream of the requested files.
+
+###Status COdes
```
Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: lower case O

File: apiservers/engine/API-COMPATIBILITY.md [hunk 11]
```diff
@@ -10,44 +76,694 @@ GET /containers/json
...
+
+This api returns a stream of the containers resource usage.
+
+__Not Supported BY VIC__
+
+the isolation unit of a containerVM is a vm so these stats are actively available from vSphere.
+
+## Resize a container TTY
+
+__Not Supported by VIC__
```
Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: Based on the hack up by George today, we're going to have this support in the tether.  We should probably add this support as it's considered a measure of quality for attach.

File: apiservers/engine/API-COMPATIBILITY.md [hunk 12]
```diff
@@ -10,44 +76,694 @@ GET /containers/json
...
+
+## Unpause a container
+
+__Not Supported By VIC__
+
+## Attach to a container
+ 
+```POST /containers/(id or name)/attach```
+
+This call will return a stream of the targeted containers common io streams(stdout, stdin, stderr). It is important to not that this involves a `HTTP UPGRADE` response.
```
Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: note spelled wrong.

File: apiservers/engine/API-COMPATIBILITY.md [hunk 13]
```diff
@@ -10,44 +76,694 @@ GET /containers/json
...
+
+* 101 : no error, indicates proxy about hijacking
+* 200 : no error, no upgrade header found
+* 400 : bad parameter
+* 404 : no such container
+* 500 : server error
+
+## Wait a container
+
+__Not Supported By VIC__
```
Comment (sflxn) [inline apiservers/engine/API-COMPATIBILITY.md]: We're probably going to have to implement this to support attach.

File: apiservers/engine/API-COMPATIBILITY.md (hunk 12 above)
Comment (matthewavery) [inline apiservers/engine/API-COMPATIBILITY.md]: Done

File: apiservers/engine/API-COMPATIBILITY.md (hunk 11 above)
Comment (matthewavery) [inline apiservers/engine/API-COMPATIBILITY.md]: Done

File: apiservers/engine/API-COMPATIBILITY.md (hunk 10 above)
Comment (matthewavery) [inline apiservers/engine/API-COMPATIBILITY.md]: Done

**Comment by sflxn** [comment]
These documents can get a little unwieldy.  The pattern I took for the CLI compatibility doc was to create a table of which CLI option we were supporting.  For operations where there are differences (parameters, json), I would write an exception entry below.  For operations that would operate just like Docker, I left it up to the reader to go to docker's website for that info.

We should remove APIs that do not have exceptions from the detailed info below.

File: apiservers/engine/API-COMPATIBILITY.md (hunk 13 above)
Comment (matthewavery) [inline apiservers/engine/API-COMPATIBILITY.md]: Done

**Comment by matthewavery** [comment]
Sounds good I will look at getting that done tonight.

**Comment by matthewavery** [comment]
@sflxn Looks like we are caught up so far.

**Comment by sflxn** [comment]
One last thing Matt, let's add a link to Docker's v1.22 API page at the top and tell the reader to go there for reference.

**Comment by matthewavery** [comment]
@loc The link has been added.

**Comment by sflxn** [comment]
lgtm
